update fetch guide

wbamberg opened the pull request:
This is a rewrite of the "Using Fetch" guide": https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API/Using_Fetch.

The main impetus here was to have a place to talk about bodies being streams, so we can document the exceptions that get thrown when a body is "locked or disturbed" (https://github.com/mdn/content/issues/13208#issuecomment-1492638882). But unfortunately the structure of this document didn't lend itself to just adding something about that, so I have restructured and rewritten it pretty much completely.

The old doc reads like a collection of facts about fetch in no particular order. In this doc I have three main parts:
- making a request, which has subtopics on the main options you can set
- cancelling a request
- handling a response, which includes checking status and reading/streaming responses

I have thrown out various things:
- "uploading a file" which I think properly belongs in the `FormData` docs (https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest_API/Using_FormData_Objects#sending_files_using_a_formdata_object , but perhaps we should have better linking to that?)
- some stuff comparing with jQuery.ajax, which I think we can dispense with now
- feature-detection stuff, likewise

Other than that I've tried to keep anything in there that seemed useful, although I've reorganized it a lot.

As well as content about streaming responses, I have added quite a bit on cross-origin requests.

I have not described all the options. Specifically I've not described the following:

- attributionReporting
- browsingTopics
- cache
- integrity
- keepalive
- priority
- redirect
- referrer
- referrerPolicy

I feel like describing all of them would make this doc a real slog to read, so I tried describing "the main ones" but that's quite subjective.

*****

I'd love feedback about whether this is a generally good "shape" of doc to have here and whether the scope is OK.

Also I this this doc needs careful review by people with really good Fetch knowledge, since some of it is tricky, and I don't know it very well.

In doing this I realize there's definite scope for more Fetch guides, like maybe separate ones of redirects, streaming, CORS, "using fetch securely", ... but I've not attempted that here. I'd welcome feedback on whether I should attempt that in this PR, especially given some of the issues in https://github.com/mdn/content/pull/34278#issuecomment-2181810190...

- Fix https://github.com/mdn/content/issues/10086
- Fix https://github.com/mdn/content/issues/24043
- Fix https://github.com/mdn/content/issues/7025

wbamberg committed:
update fetch guide

wbamberg committed:
tiny wording fix

wbamberg committed:
Remove duplicate clause

Josh-Cena commented:
Here are all issues concerning this page:

- https://github.com/mdn/content/issues/31841
- https://github.com/mdn/content/issues/24982
- https://github.com/mdn/content/issues/24043
- https://github.com/mdn/content/issues/14858
- https://github.com/mdn/content/issues/10086
- https://github.com/mdn/content/issues/7025

Maybe check if this PR happens to fix any of them, or if they are easy enough to be addressed in this PR, or if they aren't really worth fixing?

wbamberg committed:
Explain a bit about headers and bodies

wbamberg commented:
https://github.com/mdn/content/pull/34278/commits/47767e72f8ba04ca936d53d480c9dcedc2db9687 adds a bit of explanation about what headers and bodies are, in an attempt to fix https://github.com/mdn/content/issues/24043.

wbamberg committed:
Add a note about SameSite

wbamberg commented:
https://github.com/mdn/content/pull/34278/commits/404963d029edd568773e75a24b2841983d710bca notes that same-site cookies will not be sent cross-site, whatever the value of `credentials`, in an attempt to fix https://github.com/mdn/content/issues/7025.

wbamberg commented:
Thanks for this!

> * [Fetch Basic Sample Code 404 #31841](https://github.com/mdn/content/issues/31841)

As discussed in that issue, I'm open to this but there are some wrinkles and questions, so not sure if we want to do it now.

> * [Using fetch: Mixing await/async and .then/.finally syntax #24982](https://github.com/mdn/content/issues/24982)

The main bit of this is addressed. The rest seems to be about potentially splitting this guide into  "basic fetch" and separate guides for more detailed things. Again I can see the value of that and if that's the way people want to go with this PR I'm happy to. But we'd need to think of what the other things would be. Maybe:

- cross-origin requests (including CORS, credentials, CSP)
- POST requests
- redirects?
- streaming responses (but even so I think we need to talk about "disturbed" and "locked" in the basic guide, since it affects basic usage as well)

> * [Using fetch: explain more about what body/header does #24043](https://github.com/mdn/content/issues/24043)

Added https://github.com/mdn/content/commit/47767e72f8ba04ca936d53d480c9dcedc2db9687 in this PR which should address that.

> * [The Example for Form URL Encode and JSON is lost for request. New JS programmer that want to see the example can't find one #14858](https://github.com/mdn/content/issues/14858)

I wonder if this indicates that we should have a separate "Making POST requests" guide, to provide a home for this and for the "upload a file" and upload forms" examples.

> * [Issue with "Using Fetch": (makeTextFileLineIterator ignores UTF8 multi-byte boundaries) #10086](https://github.com/mdn/content/issues/10086)

I'm not at all familiar with the streams API but:
- this is already quite a complicated example in which a lot of the complexity comes out of text handling, rather than fetch. So it's not a great example of response streaming per se, and I'd be reluctant to make it more complicated (rather than change the example to a simpler use case).
- it seems weird that we have to build our own UTF8 on top of TextDecoder, which is....a UTF8 decoder. Can we use [`TextDecoderStream`](https://developer.mozilla.org/en-US/docs/Web/API/TextDecoderStream) here? Like some version of:
```js
  fetch(`/receive?channel=${channel}`).then(async res => {
    const reader = res.body.pipeThrough(new TextDecoderStream()).getReader();
    while (true) {
      const { done, value } = await reader.read();
      if (done) return;
      output.append(value);
    }
  });
```

...from https://glitch.com/edit/#!/fetch-request-stream?path=static%2Fmain.js%3A50%3A0 ?

> * [Issue with "Using Fetch": Add information about SameSite for cookies, when using credentials: 'include' #7025](https://github.com/mdn/content/issues/7025)

Added https://github.com/mdn/content/commit/404963d029edd568773e75a24b2841983d710bca in this PR which should address that.

Josh-Cena commented:
Great, so in conclusion this PR (aims to) fix #24043 and #7025?

Josh-Cena commented on files/en-us/web/api/fetch_api/using_fetch/index.md:
```diff
@@ -6,211 +6,400 @@ page-type: guide
-The [Fetch API](/en-US/docs/Web/API/Fetch_API) provides a JavaScript interface for accessing and manipulating parts of the [protocol](/en-US/docs/Glossary/Protocol), such as requests and responses. It also provides a global {{domxref("fetch()")}} method that provides an easy, logical way to fetch resources asynchronously across the network.
+The [Fetch API](/en-US/docs/Web/API/Fetch_API) provides a JavaScript interface for making HTTP requests and processing the responses.
 
-Unlike {{domxref("XMLHttpRequest")}} that is a callback-based API, Fetch is promise-based and provides a better alternative that can be easily used in [service workers](/en-US/docs/Web/API/Service_Worker_API). Fetch also integrates advanced HTTP concepts such as [CORS](/en-US/docs/Web/HTTP/CORS) and other extensions to HTTP.
```
I feel like we should give `XMLHttpRequest` _some_ attribution, because people naturally would want to compare them (you still encounter tutorials that use XHR every other day). This paragraph seems fine to me.

wbamberg replied:
Added back a slightly different version of this in https://github.com/mdn/content/pull/34278/commits/d111ac81e7d64f6148a6eaf3ddd305434858a95e (I didn't particularly like the unsupported "better alternative" judgement).

Josh-Cena commented on files/en-us/web/api/fetch_api/using_fetch/index.md:
```diff
@@ -6,211 +6,400 @@ page-type: guide
+### Setting a body
 
-## Aborting a fetch
+The request body is the payload of the request: it's the thing the client is sending to the server. You won't include a body with `GET` requests, but you will include it in requests that send content to the server, such as {{httpmethod("POST")}} or {{httpmethod("PUT")}} requests. For example, if you want to upload a file to the server, you might make a `POST` request and include the file as the r ...
```
```suggestion
The request body is the payload of the request: it's the thing the client is sending to the server. You cannot include a body with `GET` requests, but it's useful for requests that send content to the server, such as {{httpmethod("POST")}} or {{httpmethod("PUT")}} requests. For example, if you want to upload a file to the server, you might make a `POST` request and include the file as the request body.
```

("Cannot" is the important bit here)

wbamberg commented:
> Great, so in conclusion this PR (aims to) fix #24043 and #7025?

At least those two :).

Josh-Cena commented on files/en-us/web/api/fetch_api/using_fetch/index.md:
```diff
@@ -6,211 +6,400 @@ page-type: guide
+
+To control whether or not the browser sends credentials, set the `credentials` option, which can take one of the following three values:
+
+- `omit`: never send credentials or include credentials in the response.
```
```suggestion
- `omit`: never send credentials in the request and ignore credentials in the response.
```

(I was reading it as "send credentials in the response" and got confused, or maybe you really just meant "in the request", not sure)

wbamberg replied:
"ignore" feels a little vague to me, could we say:

```suggestion
- `omit`: never send credentials in the request or include credentials in the response.
```
?

Josh-Cena replied:
Sure, sounds good

Josh-Cena reviewed (commented):
Looks great!

wbamberg commented:
> > * [Issue with "Using Fetch": (makeTextFileLineIterator ignores UTF8 multi-byte boundaries) #10086](https://github.com/mdn/content/issues/10086)
> 
> I'm not at all familiar with the streams API but:
> 
> * this is already quite a complicated example in which a lot of the complexity comes out of text handling, rather than fetch. So it's not a great example of response streaming per se, and I'd be reluctant to make it more complicated (rather than change the example to a simpler use case).
> 
> * it seems weird that we have to build our own UTF8 on top of TextDecoder, which is....a UTF8 decoder. Can we use [`TextDecoderStream`](https://developer.mozilla.org/en-US/docs/Web/API/TextDecoderStream) here? Like some version of:
> 
> 
> ```js
>   fetch(`/receive?channel=${channel}`).then(async res => {
>     const reader = res.body.pipeThrough(new TextDecoderStream()).getReader();
>     while (true) {
>       const { done, value } = await reader.read();
>       if (done) return;
>       output.append(value);
>     }
>   });
> ```
> 
> ...from https://glitch.com/edit/#!/fetch-request-stream?path=static%2Fmain.js%3A50%3A0 ?
> 

So for instance, can we have a streaming example like this:

```js
const url = "https://www.gutenberg.org/cache/epub/34225/pg34225.txt";

async function fetchAsStream(url) {
  try {
    const response = await fetch(url);
    if (!response.ok) {
      throw new Error(`Response status: ${response.status}`);
    }

    const reader = response.body.pipeThrough(new TextDecoderStream());
    for await (const value of reader) {
      console.log(value);
    }
  } catch (e) {
    console.error(e);
  }
}
```

I've tried this in Chrome with network throttling, and it works nicely to deliver chunks of text as they come in (compared with the `await response.text()` version, which takes ages to fulfill). Would be nice to have a large text file that works cross-origin, which gutenberg doesn't.

I think this is a lot clearer than the "read line by line" example, and shows the point of streaming responses quite clearly. WDYT?

Josh-Cena commented:
That looks great to me 👍

Josh-Cena commented:
> I think this is a lot clearer than the "read line by line" example, and shows the point of streaming responses quite clearly. WDYT?

That looks great to me 👍

> > Great, so in conclusion this PR (aims to) fix #24043 and #7025?
> 
> At least those two :).

I've linked them to this PR

wbamberg committed:
Add a simpler streaming example;fix UTF-8 handling in the more complicated one

wbamberg commented:
This commit: https://github.com/mdn/content/pull/34278/commits/fbf188b48102f774cebdae38ca0597a0a97d5cff adds the simpler streaming example, and also fixes https://github.com/mdn/content/issues/10086 by piping the response body through a UTF-8 decoder.

I've confirmed (using https://www.gutenberg.org/cache/epub/30774/pg30774.txt) that:
- the old code did mess up multibyte UTF-8 encodings when the happened at chunk boundaries
- the new code doesn't

wbamberg committed:
Update files/en-us/web/api/fetch_api/using_fetch/index.md

Co-authored-by: Joshua Chen <sidachen2003@gmail.com>

wbamberg committed:
Reinstate para about XHR

wbamberg committed:
Delete basic concepts guide

wbamberg commented:
In https://github.com/mdn/content/pull/34278/commits/c4415968293274cceca002fd9b520274faeb1aee I deleted https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API/Basic_concepts, because it only talked about "guard", which is an internal spec concept, and whose effect is not documented in the main "using fetch" guide.

As usual I have no idea why CI is failing:

```
deletedFragmentDetails [
  'web/api/fetch_api/basic_concepts#in_a_nutshell',
  'web/api/fetch_api/basic_concepts#guard',
  'web/api/fetch_api/using_fetch#supplying_request_options',
  'web/api/fetch_api/using_fetch#aborting_a_fetch',
  'web/api/fetch_api/using_fetch#sending_a_request_with_credentials_included',
  'web/api/fetch_api/using_fetch#uploading_json_data',
  'web/api/fetch_api/using_fetch#uploading_a_file',
  'web/api/fetch_api/using_fetch#uploading_multiple_files',
  'web/api/fetch_api/using_fetch#checking_that_the_fetch_was_successful',
  'web/api/fetch_api/using_fetch#supplying_your_own_request_object',
  'web/api/fetch_api/using_fetch#headers',
  'web/api/fetch_api/using_fetch#guard',
  'web/api/fetch_api/using_fetch#response_objects',
  'web/api/fetch_api/using_fetch#body',
  'web/api/fetch_api/using_fetch#feature_detection',
  'web/api/fetch_api/using_fetch#differences_from_jquery.ajax'
]
Error: ERROR:files/en-us/web/api/fetch_api/index.md:52:37:Slug 'web/api/fetch_api/basic_concepts' has been deleted
Error: Process completed with exit code 1.
```

??

wbamberg committed:
Remove link to deleted page

wbamberg committed:
Update files/en-us/web/api/fetch_api/using_fetch/index.md

Josh-Cena commented on files/en-us/web/api/fetch_api/using_fetch/index.md:
```diff
@@ -374,118 +352,230 @@ async function fetchJSON(request) {
+
+  if (startIndex < chunk.length) {
+    // Last line didn't end in a newline char
+    yield chunk.substr(startIndex);
```
```suggestion
    yield chunk.substring(startIndex);
```

Josh-Cena reviewed (commented):
Leaving a verbal approval; if you don't get it merged in a week we can merge it

wbamberg committed:
Update files/en-us/web/api/fetch_api/using_fetch/index.md

Co-authored-by: Joshua Chen <sidachen2003@gmail.com>

Josh-Cena commented:
(@sideshowbarker bumping this up your inbox)

sideshowbarker commented on files/en-us/web/api/fetch_api/using_fetch/index.md:
```diff
@@ -6,365 +6,343 @@ page-type: guide
+The [Fetch API](/en-US/docs/Web/API/Fetch_API) provides a JavaScript interface for making HTTP requests and processing the responses.
 
-Unlike {{domxref("XMLHttpRequest")}} that is a callback-based API, Fetch is promise-based and provides a better alternative that can be easily used in [service workers](/en-US/docs/Web/API/Service_Worker_API). Fetch also integrates advanced HTTP concepts such as [CORS](/en-US/docs/Web/HTTP/CORS) and other extensions to HTTP.
+Fetch is the modern replacement {{domxref("XMLHttpRequest")}}: unlike `XMLHttpRequest`, which uses callbacks, Fetch is promise-based and is integrated with features of the modern web such as [service workers](/en-US/docs/Web/API/Service_Worker_API) and [Cross-Origin Resource Sharing (CORS)](/en-US/docs/Web/HTTP/CORS).
```
```suggestion
Fetch is the modern replacement for {{domxref("XMLHttpRequest")}}: unlike `XMLHttpRequest`, which uses callbacks, Fetch is promise-based and is integrated with features of the modern web such as [service workers](/en-US/docs/Web/API/Service_Worker_API) and [Cross-Origin Resource Sharing (CORS)](/en-US/docs/Web/HTTP/CORS).
```

sideshowbarker commented on files/en-us/web/api/fetch_api/using_fetch/index.md:
````diff
@@ -6,365 +6,343 @@ page-type: guide
 
-uploadMultiple(formData);
-```
+Including credentials in cross-origin requests can make a site vulnerable to {{glossary("CSRF")}} attacks, so even if `credentials` is set to `include`, the server must also agree to their inclusion by including the {{httpheader("Access-Control-Allow-Credentials")}} in its response. Additionally, in this situation the server must explicitly specify the client's origin in the {{httpheader("Access- ...
````
```suggestion
Including credentials in cross-origin requests can make a site vulnerable to {{glossary("CSRF")}} attacks, so even if `credentials` is set to `include`, the server must also agree to their inclusion by including the {{httpheader("Access-Control-Allow-Credentials")}} header in its response. Additionally, in this situation the server must explicitly specify the client's origin in the {{httpheader("Access-Control-Allow-Origin")}} response header (that is, `*` is not allowed).
```

sideshowbarker reviewed (approved):
Looks great to me — the technical content of the prose and examples. Only found a couple minor proofreading fixes.

wbamberg committed:
Update files/en-us/web/api/fetch_api/using_fetch/index.md

Co-authored-by: sideshowbarker <mike@w3.org>

wbamberg committed:
Update files/en-us/web/api/fetch_api/using_fetch/index.md

Co-authored-by: sideshowbarker <mike@w3.org>

wbamberg commented:
Thank you for the reviews!

wesinator commented:
I liked the example https://web.archive.org/web/20240612035627/https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API/Using_Fetch#supplying_request_options because it was a complete structure, and the other examples in the old page had examples for specific use cases where one could learn from the examples, not simply atomic pieces of the Fetch API. (I used this old example recently and found it helpful, then I noticed this documentation change).

Would it be possible to add the Supplying Request Options example back?

[NO_TRAIN]::

wbamberg commented:
Of course, but I'd like to understand better what you mean by "a complete structure". The example at the start of the new page is also "complete" in that it runs, as is the "making a POST request" example and most of the other examples in the page. On the other hand https://web.archive.org/web/20240612035627/https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API/Using_Fetch#supplying_request_options is not complete in the sense of showing all the options, it only shows some of them.

Josh-Cena commented:
-1 to add it back, or to have any example that has every option configured. If you want to read up on all possible options, read the reference instead of the guide, which shows you how to use the API。
